### PR TITLE
introduce --parser runtime flag

### PR DESCRIPTION
**Pull request by HParker**
Introduce --parser runtime flag
    
```
ruby --parser=prism
```
    
also update the description:
    
```
$ ruby --parser=prism --version
ruby 3.3.0dev (2023-12-08T04:47:14Z add-parser-runtime.. 0616384c9f) +PRISM [x86_64-darwin23]
```

[Bug #20044](https://bugs.ruby-lang.org/issues/20044)

### CODE REVIEW

**Review by kddnewton — APPROVED** [review]

File: ractor_core.h
Comment (ko1) [inline ractor_core.h:189]: I believe it should be global, not ractor local.

**Review by ko1 — CHANGES_REQUESTED** [review]

**Comment by HParker** [comment]
I have been thinking about this change and if it should go in before or after 3.3.0.

I want to have it in 3.3.0, but I worry that it might make a bad first impression for Prism. I worry that someone will turn on prism via this option and be disappointed and decide Prism is not good when they run into issues. If we get this in shortly after the 3.3.0 release and have all the time between then and 3.3.1 to test, we might be able to make a better first impression.

I am open to whatever the team thinks is best, but right now I am nervous about including it in 3.3.0.

**Comment by jeremyevans** [comment]
My opinion is that we should not add features to tiny releases, only to minor ones. So if `--parser=prism` is not ready for 3.3.0, it should wait until 3.4.0.

**Comment by HParker** [comment]
In that case, I think we should probably add this in 3.3.0. My hope is that there is a way to use prism optionally as the parser before 3.4.0. I would like to at least be testing with prism in CI regularly in 2024.

**Comment by kddnewton** [comment]
> My opinion is that we should not add features to tiny releases, only to minor ones. So if `--parser=prism` is not ready for 3.3.0, it should wait until 3.4.0.

I agree, I don't think we should add features in tiny releases. Given that we now compile all nodes and are just fixing bugs, I think we need to get this merged for 3.3.0 so that we're not stuck waiting for it for another year. The function of this flag is not for people to run it in production, the function is for us to gather feedback about what may or may not be breaking with people's code that we don't have access to. We need this flag for testing and I wouldn't want to add it in a tiny release.

**Comment by kddnewton** [comment]
@ko1 can you rereview? It looks like @HParker addressed your concern. I just want to make sure you are happy with this before I merge it. Given timezones, I'll wait for your reply tomorrow.

**Comment by kddnewton** [comment]
Sorry I just clicked the wrong button....

**Review by kddnewton — APPROVED** [review]
Looks great!

File: ruby.c
Comment (eregon) [inline ruby.c:1414]: `an issues`, should be without `an`

Comment (kddnewton) [reply ruby.c:1414]: I assume that's meant to be `any`

Comment (eregon) [reply ruby.c:1414]: Yeah `any issue` would also work

File: test/ruby/test_rubyoptions.rb
Comment (eregon) [inline test/ruby/test_rubyoptions.rb:291]: I guess this hardcoded path won't work in CI. It's likely possible to use a regexp or #include? Instead

Comment (HParker) [reply test/ruby/test_rubyoptions.rb]: Thanks yeah, updated to not include the path.

File: ruby.c
Comment (paracycle) [inline ruby.c:1414]: Can we not compare the "Prism compiler" to "parse.y" please? It is the "Prism parser" that is comparable to "parse.y", not the compiler based on that parse tree.

How does the following read?
```suggestion
            rb_warn("The compiler based on the Prism parser is currently experimental and compatibility with the current Ruby compiler is not yet complete. Please report any issues you find on the `ruby/prism` issue tracker.");
```

Comment (HParker) [reply ruby.c]: I feel like this message is a bit more complicated this way, but I agree it was inaccurate about which bits compare with each other.

File: test/ruby/test_rubyoptions.rb
Comment (paracycle) [inline test/ruby/test_rubyoptions.rb:291]: This message needs to be updated as well.

**Comment by ko1** [comment]
> @ko1 can you rereview? It looks like @HParker addressed your concern. I just want to make sure you are happy with this before I merge it. Given timezones, I'll wait for your reply tomorrow.

Sorry I missed the comment.
It's okay but `rb_ruby_` can be located on ruby.c (natural for me).

**Pull request merged**
